### DISCUSSION TITLE
How do I ScalarMul outside a circuit?

### DISCUSSION THREAD

**Discussion by Scratch-net** [Consensys/gnark#1290] [Q&A] [answered]
There's a series of computations involving EC math which we need to do both inside and outside the circuit, namely ScalarMul, assertIsOnCurve, Inverse and couple of other things needed to prove an OPRF execution for BN254 curve
However, most of them are either internal or available only inside the circuit.
Am I missing smth or the only way is to implement everything manually?

**Comment by ivokub** [comment, marked as answer]
For out-circuit computation there are corresponding methods in the `github.com/consensys/gnark-crypto` package.

I'd recommend looking at the scalar multiplication examples how the computations are performed: https://pkg.go.dev/github.com/consensys/gnark@v0.11.0/std/algebra/emulated/sw_emulated#example-Curve
